text/0000-cargo-workspace-deduplicate.md

```diff
@@ -0,0 +1,698 @@
+`false` is disallowed for now. A string can also be used like so:
+
+```toml
+[dependencies]
+log = { workspace = "log2" }
+```
+
+This directive indicates that `{ workspace = ... }` will be replaced with the
+directive in `[workspace.dependencies]`. The key to replace is either the name
+of the key in `[dependencies]` if `workspace = true` is specified, or if
+`workspace = "string"` is specified then `string` will be looked up in
+`[workspace.dependencies]`. It's an error to specify a dependency with
+`workspace` that isn't listed in `[workspace.dependencies]`.
+
+For example this directive will be replaced with the
+`workspace.dependencies.log` directive:
+
+```toml
+[dependencies]
+log = { workspace = true }
+```
+
+while this directive will be replaced with the `workspace.dependencies.log2`
+  directive:
+
+```toml
+[dependencies]
+log = { workspace = "log2" }
+```
+
+### Path dependencies infer `version` directive
+
+As a final change to `Cargo.toml`, dependencies using the `path` directive and
+not specifying a `version` directive will have the `version` directive inferred.
+
+For example if we have:
+
+```toml
+# foo/Cargo.toml
+[dependencies]
+bar = { path = "../bar" }
+```
+
+as well as
+
+```toml
+# bar/Cargo.toml
+[package]
+name = "bar"
+version = "1.0.1"
+```
+
+this is equivalent in `foo/Cargo.toml` to as if this were written:
+
+```toml
+# foo/Cargo.toml
+[dependencies]
+bar = { path = "../bar", version = "1.0.1" }
+```
+
+The `version` key for `path` dependencies, if not specified, will be inferred to
+the version of the path dependency itself. Note that this is a version
+requirement not an actual semver vesion, and the version requirement will be
+interpreted as "at least the current version, and anything semver compatible
+with it".
+
+This logic of inferring, however, will also respect the `publish` key. For
+example if we had this instead:
+
+```toml
+# bar/Cargo.toml
+[package]
+name = "bar"
+version = "1.0.1"
+publish = false
+```
+
+then Cargo would not alter this dependency directive:
+
+```toml
+# foo/Cargo.toml
+[dependencies]
+bar = { path = "../bar" }
+```
+
+## Effect on `cargo publish`
+
+Cargo currently already "elaborates" the manifest during publication. For
+example it removes `path` keys in dependency lists to only have the version
+requirement pointing to crates.io. During publication Cargo will also elaborate
+any substituted information from the `[workspace]`, because `[workspace]` is
+also removed during publication!
+
+This means that `workspace = true` will never be present in `Cargo.toml` files
+published to crates.io, and additionally no information about `workspace = true`
+will make its way to the registry index. Furthermore metadata fields like
+`package.repository` will be filled in and will be present on crates.io's UI.
+
+Put another way, `Cargo.toml` files published to crates.io, or metadata found
+through crates.io, won't change from what they are today.
```

Comment:
Overall, this is a great RFC that addresses some papercuts I've seen in https://github.com/tokio-rs/tracing. I do have a question: does this RFC make a `cargo publish --all` (that publishes all unpublished crates in a workspace) easier to support or is RFC orthogonal to that feature?

Reply:
I don't think this makes that either harder or easier per se, since this is, internally in Cargo at least, mostly just "syntax sugar". I would guess that an implementation of `publish --all` wouldn't have to specifically handle anything in this rfc (in that it doesn't make it harder), and it would also have to handle everything still (this doesn't make it easier)

Reply:
@davidbarsky I implemented a nice way of doing that (using dep graph of only the crates you are publishing) in https://github.com/pksunkara/cargo-workspaces